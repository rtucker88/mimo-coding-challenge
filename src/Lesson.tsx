import * as classnames from 'classnames';
import { flatten, isUndefined } from 'lodash';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { StateHandler, StateHandlerMap, withStateHandlers } from 'recompose';
import { ILesson } from './services/lesson';

import './lesson.css';

interface ILessonProps {
    lesson: ILesson;
    nextLesson: number | 'DONE';
    completeLesson: () => void;
}

interface ILessonState {
    isLessonFinished: boolean;
}

interface ILessonStateHandlers extends StateHandlerMap<ILessonState> {
    onInputChange: StateHandler<ILessonState>;
}

export const enhance = withStateHandlers<ILessonState, ILessonStateHandlers, ILessonProps>((props) => ({
    isLessonFinished: isUndefined(props.lesson.input),
}), {
    onInputChange: (_, props) => (event: React.ChangeEvent<HTMLInputElement>) => {
        return {
            isLessonFinished: getSolutionValue(props.lesson) === event.target.value
        };
    }
});

export type CombinedLessonProps = ILessonProps & ILessonState & ILessonStateHandlers;

const Lesson: React.StatelessComponent<CombinedLessonProps> = ({ lesson, nextLesson, completeLesson, onInputChange, isLessonFinished }) => {
    return (<div>
        {createLesson(lesson, onInputChange)}
        <Link to={`/${nextLesson}`} className={classnames({'disabled-link': !isLessonFinished})}><button disabled={!isLessonFinished} onClick={completeLesson}>Continue</button></Link>
    </div>);
};

export default enhance(Lesson);

export function getSolutionValue({ content, input }: ILesson): string {
    return content.map(cnt => cnt.text).join('').substr(input!.startIndex, input!.endIndex - input!.startIndex);
}

// Assuming valid data when going through this process. The error cases quickly result in a broken UI and 
// I'm unsure how we would recover from that state
function createLesson(lesson: ILesson, onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => Partial<ILessonState> | undefined) {
    let textCounter = 0; // Use this to figure out where we replace text with an input
    
    const elems = lesson.content.map(entry => {
        const style: React.CSSProperties = {color: entry.color};
        const retElems: JSX.Element[] = [];

        const startIndex = textCounter;
        const endIndex = textCounter + entry.text.length;

        // Tokenize around the input
        if (lesson.input && lesson.input.startIndex >= startIndex && lesson.input.endIndex < endIndex) {
            // Before Input
            retElems.push(<span style={style} key={`${textCounter}-before`}>{entry.text.substr(0, lesson.input.startIndex - startIndex)}</span>)
            // Input
            retElems.push(<input style={style} key={textCounter} maxLength={lesson.input.endIndex - lesson.input.startIndex} onChange={onInputChange} />);
            // After Input
            retElems.push(<span style={style} key={`${textCounter}-after`}>{entry.text.substr(lesson.input.endIndex - startIndex, entry.text.length)}</span>)
        } else {
            retElems.push(<span style={style} key={textCounter}>{entry.text}</span>)
        }
        
        textCounter += entry.text.length;

        return retElems;
    });

    return flatten(elems);
}