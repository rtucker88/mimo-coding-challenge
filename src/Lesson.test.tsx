import { mount } from 'enzyme';
import * as React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import Lesson, { CombinedLessonProps, enhance } from './Lesson';

describe('Lesson', () => {
    class MockComponent extends React.Component<CombinedLessonProps> {
        public render() {
            return (<div/>);
        }
    }

    const lesson = {
        content: [{
            color: '#FFF',
            text: 'var number =',
        }],
        id: 1,
    };

    const lessonWithInput = {
        content: [{
            color: '#FFF',
            text: 'var number =',
        }],
        id: 1,
        input: {
            endIndex: 2,
            startIndex: 1,
        }
    };

    const completeLessonMock = jest.fn();

    beforeEach(() => {
        completeLessonMock.mockReset();
    });

    describe('renders correctly', () => {
        it('without inputs', () => {
            const wrapper = mount((
                <Router>
                    <Lesson lesson={lesson} nextLesson={2} completeLesson={completeLessonMock} />
                </Router>)
            );
            expect(wrapper).toMatchSnapshot();
        });
        
        it('with inputs', () => {
            const wrapper = mount((
                <Router>
                    <Lesson lesson={lessonWithInput} nextLesson={2} completeLesson={completeLessonMock} />
                </Router>)
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    it('handles an correct input change correctly', () => {
        const Enhanced = enhance(MockComponent);
        const wrapper = mount(<Enhanced lesson={lessonWithInput} nextLesson={2} completeLesson={completeLessonMock} />);
        expect(wrapper.state('isLessonFinished')).toEqual(false);

        const subject = wrapper.find(MockComponent);
        subject.prop('onInputChange')({ target: { value: 'a' }});
    
        expect(wrapper.state('isLessonFinished')).toEqual(true);
    });

    it('handles an incorrect input change correctly', () => {
        const Enhanced = enhance(MockComponent);
        const wrapper = mount(<Enhanced lesson={lessonWithInput} nextLesson={2} completeLesson={completeLessonMock} />);
        expect(wrapper.state('isLessonFinished')).toEqual(false);

        const subject = wrapper.find(MockComponent);
        subject.prop('onInputChange')({ target: { value: 'b' }});
    
        expect(wrapper.state('isLessonFinished')).toEqual(false);
    });

    it('should always have a valid solution value if there is no input', () => {
        const Enhanced = enhance(MockComponent);
        const wrapper = mount(<Enhanced lesson={lesson} nextLesson={2} completeLesson={completeLessonMock} />);
        expect(wrapper.state('isLessonFinished')).toEqual(true);
    });
});
