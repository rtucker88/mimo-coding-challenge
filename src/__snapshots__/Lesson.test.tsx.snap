// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson renders correctly with inputs 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withStateHandlers(Lesson)
      completeLesson={[MockFunction]}
      lesson={
        Object {
          "content": Array [
            Object {
              "color": "#FFF",
              "text": "var number =",
            },
          ],
          "id": 1,
          "input": Object {
            "endIndex": 2,
            "startIndex": 1,
          },
        }
      }
      nextLesson={2}
    >
      <Lesson
        completeLesson={[MockFunction]}
        isLessonFinished={false}
        lesson={
          Object {
            "content": Array [
              Object {
                "color": "#FFF",
                "text": "var number =",
              },
            ],
            "id": 1,
            "input": Object {
              "endIndex": 2,
              "startIndex": 1,
            },
          }
        }
        nextLesson={2}
        onInputChange={[Function]}
      >
        <div>
          <span
            key="0-before"
            style={
              Object {
                "color": "#FFF",
              }
            }
          >
            v
          </span>
          <input
            key="0"
            maxLength={1}
            onChange={[Function]}
            style={
              Object {
                "color": "#FFF",
              }
            }
          />
          <span
            key="0-after"
            style={
              Object {
                "color": "#FFF",
              }
            }
          >
            r number =
          </span>
          <Link
            className="disabled-link"
            replace={false}
            to="/2"
          >
            <a
              className="disabled-link"
              href="/2"
              onClick={[Function]}
            >
              <button
                disabled={true}
                onClick={[MockFunction]}
              >
                Continue
              </button>
            </a>
          </Link>
        </div>
      </Lesson>
    </withStateHandlers(Lesson)>
  </Router>
</BrowserRouter>
`;

exports[`Lesson renders correctly without inputs 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withStateHandlers(Lesson)
      completeLesson={[MockFunction]}
      lesson={
        Object {
          "content": Array [
            Object {
              "color": "#FFF",
              "text": "var number =",
            },
          ],
          "id": 1,
        }
      }
      nextLesson={2}
    >
      <Lesson
        completeLesson={[MockFunction]}
        isLessonFinished={true}
        lesson={
          Object {
            "content": Array [
              Object {
                "color": "#FFF",
                "text": "var number =",
              },
            ],
            "id": 1,
          }
        }
        nextLesson={2}
        onInputChange={[Function]}
      >
        <div>
          <span
            key="0"
            style={
              Object {
                "color": "#FFF",
              }
            }
          >
            var number =
          </span>
          <Link
            className=""
            replace={false}
            to="/2"
          >
            <a
              className=""
              href="/2"
              onClick={[Function]}
            >
              <button
                disabled={false}
                onClick={[MockFunction]}
              >
                Continue
              </button>
            </a>
          </Link>
        </div>
      </Lesson>
    </withStateHandlers(Lesson)>
  </Router>
</BrowserRouter>
`;
